source/dol/bases/d_mj2d_data.cpp

```diff
@@ -0,0 +1,312 @@
+}
+
+int dMj2dGame_c::getCourseDataFlag(int world, int level) const {
+    return mStageCompletion[world][level];
+}
+
+bool dMj2dGame_c::isCourseDataFlag(int world, int level, ulong flag) const {
+    return (flag & mStageCompletion[world][level]) != 0;
+}
+
+void dMj2dGame_c::onCourseDataFlag(int world, int level, ulong flag) {
+    mStageCompletion[world][level] |= flag;
+}
+
+void dMj2dGame_c::offCourseDataFlag(int world, int level, ulong flag) {
+    mStageCompletion[world][level] &= ~flag;
+}
+
+void dMj2dGame_c::setEnemyRevivalCount(int world, int enemy, u8 count) {
+    mEnemyRevivalCount[world][enemy] = count;
+}
+
+u8 dMj2dGame_c::getEnemyRevivalCount(int world, int enemy) const {
+    return mEnemyRevivalCount[world][enemy];
+}
+
+void dMj2dGame_c::setEnemySubWorld(int world, int enemy, u8 subworld) {
+    mEnemySubWorld[world][enemy] = subworld;
+}
+
+u8 dMj2dGame_c::getEnemySubWorld(int world, int enemy) const {
+    return mEnemySubWorld[world][enemy];
+}
+
+void dMj2dGame_c::setEnemyCurrNode(int world, int enemy, u8 node) {
+    mEnemyPathNode[world][enemy] = node;
+}
+
+u8 dMj2dGame_c::getEnemyCurrNode(int world, int enemy) const {
+    return mEnemyPathNode[world][enemy];
+}
+
+void dMj2dGame_c::setEnemyLastDirection(int world, int enemy, u8 direction) {
+    mEnemyDirection[world][enemy] = direction;
+}
+
+u8 dMj2dGame_c::getEnemyLastDirection(int world, int enemy) const {
+    return mEnemyDirection[world][enemy];
+}
+
+void dMj2dGame_c::setKinopioCourseNo(int world, int level) {
+    mKinopioCourseNo[world] = level;
+}
+
+u8 dMj2dGame_c::getKinopioCourseNo(int world) const {
+    return mKinopioCourseNo[world];
+}
+
+void dMj2dGame_c::setIbaraNow(int ibaraNow) {
+    mIbaraNow = ibaraNow;
+}
+
+u8 dMj2dGame_c::getIbaraNow() const {
+    return mIbaraNow;
+}
+
+dMj2dHeader_c::dMj2dHeader_c() {
+}
+
+void dMj2dHeader_c::initialize() {
+    memset(this, 0, sizeof(dMj2dHeader_c));
+
+    mMagic[0] = sSaveMagic[0];
+    mMagic[1] = sSaveMagic[1];
+    mMagic[2] = sSaveMagic[2];
+    mMagic[3] = sSaveMagic[3];
+
+    mVersion[0] = SAVE_VERSION;
+    mVersion[1] = SAVE_SUB_VERSION;
+}
+
+void dMj2dHeader_c::setSelectFileNo(s8 fileNum) {
+    mLastSelectedFile = fileNum;
+}
+
+u16 dMj2dHeader_c::getFreeModePlayCount(int world, int level) const {
+    return mFreeModePlayCount[world][level];
+}
+
+void dMj2dHeader_c::setFreeModePlayCount(int world, int level, u16 count) {
+    mFreeModePlayCount[world][level] = count;
+}
+
+u16 dMj2dHeader_c::getCoinBattlePlayCount(int world, int level) const {
+    return mCoinBattlePlayCount[world][level];
+}
+
+void dMj2dHeader_c::setCoinBattlePlayCount(int world, int level, u16 count) {
+    mCoinBattlePlayCount[world][level] = count;
+}
+
+void dMj2dHeader_c::setMultiModeCompletion(int world) {
+    mExtraModesUnlockedWorlds |= 1 << world;
+}
+
+bool dMj2dHeader_c::getMultiModeCompletion(int world) const {
+    return (mExtraModesUnlockedWorlds & 1 << world);
```

Comment:
```suggestion
    return mExtraModesUnlockedWorlds & (1 << world);
```
for better readability